subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -1655,20 +1808,92 @@ public List<String> getResolutionAlternatives() {
     }
 
     @Override
-    public DeprecatableConfiguration deprecateForDeclaration(String... alternativesForDeclaring) {
+    public void setDeprecatedForConsumption(boolean deprecated) {
+        if (consumptionDeprecated != deprecated) {
+            if (deprecated) {
+                deprecateForConsumption(depSpec -> DeprecationLogger.deprecateConfiguration(getName())
+                        .forConsumption()
+                        .willBecomeAnErrorInGradle9()
+                        .withUserManual("dependencies_should_no_longer_be_declared_using_the_compile_and_runtime_configurations"));
+            } else {
+                undeprecateForConsumption();
+            }
+        }
+    }
+
+    @Override
+    public void setDeprecatedForResolution(boolean deprecated) {
+        if (resolutionDeprecated != deprecated) {
+            if (deprecated) {
+                deprecateForResolution();
+            } else {
+                undeprecateForResolution();
+            }
+        }
+    }
+
+    @Override
+    public void setDeprecatedForDeclarationAgainst(boolean deprecated) {
+        if (declarationDeprecated != deprecated) {
+            if (deprecated) {
+                deprecateForDeclarationAgainst();
+            } else {
+                undeprecateForDeclarationAgainst();
+            }
+        }
+    }
+
+    private void undeprecateForConsumption() {
+        validateMutation(MutationType.USAGE);
+        consumptionDeprecation = null;
+        consumptionDeprecated = false;
+        logChangingUsage("deprecated for consumption", false);
+    }
+
+    private void undeprecateForResolution() {
+        validateMutation(MutationType.USAGE);
+        resolutionAlternatives = null;
+        resolutionDeprecated = false;
+        logChangingUsage("deprecated for resolution", false);
+    }
+
+    private void undeprecateForDeclarationAgainst() {
+        validateMutation(MutationType.USAGE);
+        declarationAlternatives = null;
+        declarationDeprecated = false;
+        logChangingUsage("deprecated for declaration against", false);
+    }
+
+    @Override
+    public DeprecatableConfiguration deprecateForDeclarationAgainst(String... alternativesForDeclaring) {
```

Comment:
Seems odd we have two "public/internal" versions of this method, one defined in `DeprecatableConfiguration` which accepts declaration alternatives, and the other in `ConfigurationInternal` which takes a boolean arg of whether to deprecate or not. 

We should choose one to make visible outside of this class and remove the other (if we need either at this point). Given all the changes in this PR, that might just mean we get rid of `DeprecatableConfiguration` altogether now, or we move some of the new methods added to `ConfigurationInternal` into `DeprecatableConfiguration`

Reply:
I think we need the boolean-valued methods to make the deprecated roles work.  The requirements for calling the existing methods in `DeprecatedConfiguration` are too strict, IMO.  Why should I need to tell you an alternative or provide a function to deprecate this usage, I'm being a good citizen by deprecating it in the first place, don't make the process of doing this harder!

This PR didn't aim to remove or correct the existing behavior or internal API, so I think removing `DeprecatedConfiguration` is out of scope.  I also think we need a way to undeprecate - in the same way we can toggle between allowed/unallowed usage.  The `setDeprecatedForXYZ(boolean)` methods provide this in a way that wasn't present before.

Removing the `DC` type in a separate PR is definitely something to explore afterwards.  I question how useful providing those alternatives when deprecating actually are, and would favor replacing them with a separate `setDeprecationReason(String reason)` to record a static error message.  At this point a `DefaultConfiguration` is deprecatable, and keepign these as 2 separate types doesn't make sense.

Reply:
I question the usefulness of undeprecating a configuration. Needing to un-deprecate something IMO points to a missing API for creating the deprecated configuration in the first place. The deprecated configurations are an internal thing anyways, so IMO instead of having an API to un-deprecate something, we should instead have a more rich api for creating it in the proper deprecated state in the first place.

Reply:
I saw undeprecating as a way station on the road to correct usage:

1) Gradle (or plugin B) creates a configuration with a role deprecating a usage.
2) My plugin A still requires that usage.
3) I undeprecate the usage to avoid seeing warnings.
4) Eventually, I fix my plugin.

Reply:
Couldn't I just create it in a fashion where its not deprecated to start with? We should be providing factory methods to users for each use case where one would want. Those use cases are bucket, resolvable, consumable, and "legacy" (resolvable/bucket). Any other use case is really wrong and you shouldn't be able to durn off deprecation warnings for that.

Reply:
I'm thinking of your plugin using a configuration it didn't create here.  Maybe one of the configurations automatically created for each SourceSet.

Reply:
The whole point of this effort here is to get people to stop using these configurations incorrectly. If they are using incorrectly, they should hit a deprecation warning. I'm not sure of anywhere else we're allowing users to turn off deprecation warnings. The point of the deprecation warning is to get the user to stop doing the thing they're doing. 

Anyways, are we planning to expose these un-deprecate methods to users? If so, the whole deprecation framework is going to go away in 9.0. Are we going to introduce these methods as already deprecated? Incubating? 

What happens at 9.0 if someone disables the deprecation for a configuration? Otherwise, they'd be at the last 8.x and they'd have no deprecation warnings. Upon 9.0 when there is no concept of a deprecated configuration, what happens? We'd need to eventually before 9.0 deprecate undeprecating a configuration.

Reply:
These shouldn't ever become public, they're internal, so I don't think there's a need to incubate or deprecate.  In Gradle 9.0, the goal was strict mode where usage is immediately locked to the given role on creation, so we'll remove them then.  They're just a temporary escape hatch if deprecation gets to be too painful.

It's not just for external users, but also in case plugins used in the gradle/gradle build can't be updated quickly enough.  If 99% of the time when we create configurations from a sourceset we know a usage should be deprecated, but in one particular project one particular plugin really makes heavy use of it we can temporarily undeprecate _just that configuration_ while leaving the role in place for the general case until we can upgrade the plugin.

Reply:
So you're saying we'd use these methods internally to un-deprecate certain roles of our own configurations for use within our own internal plugins? Do you have any examples of where we would need this? 

All of our major java configurations are already being created with an INTENDED role, so nothing is deprecated for them. Which configurations are we creating currently that have DEPRECATED roles that we'd potentially need to un-deprecate? 

It just seems like we're writing functionality for a problem we don't have yet, and if we do have that problem in the future, this is just a stop-gap so we can get rid of that improper configuration usage ASAP? 

It still seems like if this is all for internal use, then we should just be updating the original configuration to be deprecated for that usage. We have control over that configuration so we can change it to be un-deprecated upon construction.

Reply:
> Do you have any examples of where we would need this?

Nothing presently, but I haven't done any sort of exhaustive searching.  The JMH case just makes me think it will come up and we'll find configurations we can't fully change to `INTENDED` but need to leave `DEPRECATED`.

> It just seems like we're writing functionality for a problem we don't have yet, and if we do have that problem in the future, this is just a stop-gap so we can get rid of that improper configuration usage ASAP?

Well, yes, exactly - but when you say it like that then it sounds like a bad thing. 😀 

> We have control over that configuration so we can change it to be un-deprecated upon construction.

I think there's still a benefit to updating a reused creation pattern so that it deprecates a configuration (as intended) almost all the time, yet allows for a specific exceptional case to undo this, rather than _not deprecating_ because that specific case is too much to deal with ATM.

Reply:
I understand there may be some odd cases here and there, but why do we need to make the configurations' deprecations mutable when we can do it at creation time -- especially if this is all getting used internally? There are ways to create ad-hoc roles and create configurations from ad-hoc roles. That should be enough for these use-cases, no?

Reply:
I don't think so; I think an ad-hoc role and undeprecating are slightly different use cases.  An ad-hoc role covers _unusual_ but _consistent_ behavior (say, consumable and declarable) when you have control of configuration creation.  Un-deprecating is for when (unusual or not) behavior role is _inconsistent_ and you _don't_ have control of configuration creation.  

It's another stepping stone to allow a gradual migration to fully `INTENDED` roles that are locked at creation.

1. fully mutable allowed usage without roles
2. fully mutable allowed usage using the LEGACY role
3. fully mutable allowed usage using a DEPRECATED or adhoc role
4. fully mutable allowed usage using an INTENDED role
5. immutable allowed usage using an INTENDED role

We can set `warnOnUsageChanging` and undeprecating as tools to help move farther down that progression without edge cases blocking us.  Undeprecating handles the case when we want to move to (3) and it's fine almost everywhere except for an edge case; or when it's fine for 9/10 subprojects but not the last one.  I think it's worth building flexible tools to cover these edge cases before they come up, similar to how we can also lock individual configurations or new configurations without having to make everything strict all at once.  

Because this is a temporary internal feature with a strictly limited lifespan meant to help with a migration that we should expect will have some tricky parts, the usual objections to overbuilding features that won't be needed shouldn't be weighed quite as heavily here - we should instead provide more and finer-grained control surfaces from the start.

Reply:
> Un-deprecating is for when (unusual or not) behavior role is _inconsistent_ and you _don't_ have control of configuration creation.

But un-deprecating is an internal-only feature so we are the only ones using it and we always have control over configuration creation for our own configurations. When do our own plugins ever operate on user-created configurations which we don't control? Or do you see us selectively un-deprecating certain configurations only when certain plugins are applied? I do not think we should ever be doing that. 

> We can set `warnOnUsageChanging`

I kinda figured we would be setting this for all configurations once we got to 8.1. We want to let users be warned about configurations' roles changing as early as possible in the release cycle so they have time to update by 9.0. Is this something you think we'd be setting on a configuration-by-configuration basis? 

> Because this is a temporary internal feature with a strictly limited lifespan meant to help with a migration that we should expect will have some tricky parts

Because of this I'm going to stop arguing my case, though I'll cease with a final word that I think this is pretty smelly. I find its odd that we're designing this framework to customize configurations upon creation just to go and skip over all that work to make the deprecations mutable after the fact. I struggle to see a use-case for these additional methods that wouldn't also be very questionable.

Reply:
Reading through this again, I believe we're in agreement about the facts of this change, but are each valuing the tradeoffs and consequences differently.  Let's see what the rest of the team thinks here.

Reply:
I really don't understand the purpose of un-deprecating _as an internal feature_ after reading through this. Internal means "only projects from Gradle itself should use it", and we should be spending our time fixing it, not just slapping on a patch. Do we have any builds that gate against deprecation warnings like these? `gradle/gradle` itself doesn't, `master` has deprecation warnings right now. If not, I really don't see the purpose to this.

Reply:
> I really don't understand the purpose of un-deprecating as an internal feature after reading through this. Internal means "only projects from Gradle itself should use it", and we should be spending our time fixing it, not just slapping on a patch.

Hopefully this is the case, any issues can be quickly fixed correctly, and undeprecating isn't ever used.  I just don't see the harm in providing a fuller toolkit by making deprecation toggle-able instead of purely a one way switch in this particular scenario.

Reply:
I would rather add it in later, if we need it.

Reply:
After reading through this, I'm in agreement that I don't find the argument for undeprecation compelling.  I think there's enough complexity in what we're doing here, that we should avoid adding any more unless we absolutely need it.  If we had a concrete example of where we'll need this, it might begin to tip the argument another way, but I'm not convinced it's worth it yet.  Is there any intrinsic value to adding it now versus later?  If not, my vote is to keep this as simple as possible and only introduce undeprecation if we come across a use case where we can't move forward without it.

Reply:
OK, I surrender.  I'll remove the undeprecation.